elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java

```diff
@@ -192,81 +235,133 @@ public void fillDigest(MessageDigest digest) throws IOException {
 			digestWriter.insertNewline();
 		}
 		//Now we finally know the content size in bytes:
-		this.contentLength = digestWriter.getContentLenght();
+		hintContentLenght( digestWriter.getContentLenght() );
+	}
+
+	private void hintContentLenght(long contentLenght) {
+		if ( contentlengthWasProvided == false ) {
+			this.contentLength = contentLenght;
+		}
 	}
 
-	private static final class ProgressiveByteBufWriter extends Writer {
+	private static final class ProgressiveCharBufferWriter extends Writer {
 
-		//Initially null: must be set before writing,
-		//as it might change between writes.
+		//Initially null: must be set before writing is starter and each
+		//time it's resumed as it might change between writes during
+		//chunked encoding.
 		ContentEncoder out;
 
 		//This is to keep the buffers which could not be written to socket
 		//because of flow control. Make sure to eventually push them out.
 		//Initialised only if/when needed.
-		List<ByteBuffer> unwrittenLazyBuffers;
+		List<CharBuffer> unwrittenLazyBuffers;
 
 		//Set this to true when we detect clogging, so we can stop trying.
 		//Make sure to reset this when the HTTP Client hints so.
 		//It's never dangerous to re-enable, just not efficient to try writing
 		//unnecessarily.
 		boolean flowControlPushingBack = false;
 
-		//Accumulate the content length while it's being produced so we can hint
-		//other components.
-		long totalWrittenBytes = 0;
+		private final CharsetEncoder utf8Encoder = StandardCharsets.UTF_8.newEncoder();
+		private ByteBuffer availableBuffer = ByteBuffer.allocate( BUFFER_SIZES );
+		private ByteBuffer needsWritingBuffer;
 
 		@Override
 		public void write(char[] cbuf, int off, int len) throws IOException {
 			CharBuffer toWriteChars = CharBuffer.wrap( cbuf, off, len );
-			ByteBuffer toWriteBytes = StandardCharsets.UTF_8.encode( toWriteChars );
-			writeOrBufferWrite( toWriteBytes );
-		}
-
-		public long getTotalWrittenAndReset() {
-			final long total = totalWrittenBytes;
-			//To not count size multiple times on repeated renders:
-			totalWrittenBytes = 0;
-			return total;
+			if ( flowControlPushingBack == false ) {
+				writeOrStore( toWriteChars, false, true );
+			}
+			else {
+				storeForLater( toWriteChars );
+			}
 		}
 
 		public void resumePendingWrites() throws IOException {
 			flowControlPushingBack = false;
-			flushPendingBuffers();
+			if ( availableBuffer == null ) {
+				attemptFlushPendingBuffers( false );
+			}
 		}
 
-		public void insertNewline() throws IOException {
-			writeOrBufferWrite( ByteBuffer.wrap( NEWLINE ) );
+		private void storeForLater(CharBuffer toWriteLater) {
+			if ( unwrittenLazyBuffers == null ) {
+				unwrittenLazyBuffers = new ArrayList<>( 16 );
+			}
+			//The original CharBuffer is not safe as it's backed by a char array
+			//which happens to be the GSON write buffer: it will get mutated.
+			CharBuffer defensiveCopy = CharBuffer.allocate( toWriteLater.remaining() );
```

Comment:
It may be silly, but... since you have to perform an allocation anyway, why not directly allocate ByteBuffers, and directly encode the characters in there? You could then remove this list of char buffers, and use a list of ByteBuffers instead. This would mean turning "needsWritingBuffer" into a list, if you will: it would initially only contain one buffer, and would grow as necessary.

I tested it a bit with the performance tests, and it seems having a lot of "unwritten lazy buffers" is not rare; I was above the 16 buffer initial capacity about 25% of the time, I think, going all the way up to more than 50 buffers. So it could be useful.
But more importantly, I wonder if it couldn't make the code a bit simpler. Do you think it could?